File: tests/test_distance.py
import unittest
from distance import HammingDistance, LevenshteinDistance, ShortestDetour


class DistanceTest(unittest.TestCase):

    def test_hamming_distance(self):
        dis = HammingDistance("AAATAATCCGAG", "AAATAATCCGAG")
        self.assertEqual(dis.score, 0)
        self.assertEqual(len(dis.matrix), 13)
        self.assertEqual(len(dis.matrix[0]), 13)
        dis = HammingDistance("AAATAATGCGAG", "AAATAATCCGAG")
        self.assertEqual(dis.score, 1)

        with self.assertRaises(ValueError):
            dis = HammingDistance("AAATAATGCGAG", "AAATAATCGAG")
        
        dis = HammingDistance("AAATAATCGAGA", "AAATAATCCGAG")
        self.assertEqual(dis.score, 4)
    
    def test_levenshtein_distance(self):
        dis = LevenshteinDistance("AAATAATCCGAG", "AAATAATCCGAG")
        self.assertEqual(dis.score, 0)

        dis = LevenshteinDistance("AAATAATGCGAG", "AAATAATCCGAG")
        self.assertEqual(dis.score, 1)

        dis = LevenshteinDistance("AAATAATCGAG", "AAATAATCCGAG")
        self.assertEqual(dis.score, 1)

        dis = LevenshteinDistance("AAATAATCCGAG", "AAATAATCGAG")
        self.assertEqual(dis.score, 1)

    def test_shortest_detour(self):
        dis = ShortestDetour("AAATAATCCGAG", "AAATAATCCGAG", 5)
        self.assertEqual(dis.score, 0)

        dis = ShortestDetour("AAATAATGCGAG", "AAATAATCCGAG", 5)
        self.assertEqual(dis.score, 1)

        dis = ShortestDetour("AAATAATCGAG", "AAATAATCCGAG", 5)
        self.assertEqual(dis.score, 1)

        with self.assertRaises(ValueError):
            dis = ShortestDetour("AAATAATCCGAG", "AAATAATGAG", 5)
        
        with self.assertRaises(ValueError):
            dis = ShortestDetour("AAATAATCCGAG", "AAATAATCCGAGTCGAGC", 5)
        

File: cpp/distance/tests/test_distance.cpp
#define CATCH_CONFIG_MAIN

#include "../../catch/catch.hpp"
#include "../EditDistance.h"
#include "../HammingDistance.h"
#include "../LevenshteinDistance.h"
#include <iostream>
#include <string>
#include <limits>
#include <exception>

TEST_CASE("Testing edit distance", "[EditDistance]") {
    SECTION("Testing two equal strings") {
        const double INF = std::numeric_limits<double>::infinity();
        const std::string x = "AAATAATCCGAG";
        const std::string y = "AAATAATCCGAG";
        EditDistance d (x, y, INF, 0, 1);
        REQUIRE(d.getScore() == 0);
    }
}

TEST_CASE("Testing Hamming distance", "[HammingDistance]") {
    
    SECTION("Testing two equal strings") {
        const std::string x = "AAATAATCCGAG";
        const std::string y = "AAATAATCCGAG";
        HammingDistance d (x, y);
        REQUIRE(d.getScore() == 0);
        auto matrix = d.getMatrix();
        REQUIRE(matrix->size() == 13);
        REQUIRE((*matrix)[0].size() == 13);
    }

    SECTION("Testing two strings unequally long") {
        const std::string x = "AAATAATCCGA";
        const std::string y = "AAATAATCCGAG";
        REQUIRE_THROWS_AS([&](){HammingDistance d (x,y);}(), std::invalid_argument); 
    }
    
    SECTION("Testing different strings") {
        const std::string x = "AAATAATCGAGA";
        const std::string y = "AAATAATCCGAG";
        HammingDistance d (x, y);
        REQUIRE(d.getScore() == 4);
    }
}

TEST_CASE("Testing Levenshtein distance", "[LevenshteinDistance]") {
    
    SECTION("Testing two equal strings") {
        const std::string x = "AAATAATCCGAG";
        const std::string y = "AAATAATCCGAG";
        LevenshteinDistance d (x, y);
        REQUIRE(d.getScore() == 0);
        auto matrix = d.getMatrix();
        REQUIRE(matrix->size() == 13);
        REQUIRE((*matrix)[0].size() == 13);
    }
    
    SECTION("Testing mismatch") {
        const std::string x = "AAATAATGCGAG";
        const std::string y = "AAATAATCCGAG";
        LevenshteinDistance d (x, y);
        REQUIRE(d.getScore() == 1);
    }

    SECTION("Testing indel") {
        const std::string x = "AAATAATCGAG";
        const std::string y = "AAATAATCCGAG";
        LevenshteinDistance d (x, y);
        REQUIRE(d.getScore() == 1);
    }
}